### ISSUE TITLE
Error »Cannot find module "api/collections"« after step 4

### ISSUE THREAD

**Issue by sigalor** [Urigo/Ionic2CLI-Meteor-WhatsApp#109]
Hi,

first, thanks a lot for the amazing tutorial. I am currently trying to follow it, but failed after finishing step 4.
I have created a [repository](https://github.com/sigalor/ionic-meteor-whatsapp-error) one should be able to reproduce the error with.
When I try to run `ionic serve`, the compilation works without any errors.
But right after my browser opens the page, Ionic just says `Runtime Error: Cannot find module "api/collections"`.
Though I have very limited knowledge in Ionic/Meteor/Angular/Webpack/etc., I came to the conclusion that no files in the `api` directory are accessible by the script in `src/pages/chats/chats.ts`.
Additionally, to simplify debugging, I added the line `export const Hello = "Hello world!"` in `api/server/collections/chats.ts` and referenced it in the `chats` page.

So, is this a bug in Webpack (maybe an issue because of the path alias) or something else?
Please have in mind that I am very unexperienced in all of these modern web application libraries, so excuse if this is a typical beginner's question.

My setup:
OS: Ubuntu 16.04, 64 bit
Browser: Vivaldi 1.10.867.48, 64 bit (but it doesn't work in Chromium 59.0.3071.109 or Firefox 54.0 either)
What Ionic says at the end of its error log:
Ionic Framework: 3.5.3
Ionic App Scripts: 2.0.2
Angular Core: 4.1.3
Angular Compiler CLI: 4.1.3
Node: 6.11.0
OS Platform: Linux 4.4
Navigator Platform: Linux x86_64

**Comment by clariontools** [comment]
First double check your tsconfig.json file in the main folder against the recommended configuration, then do  the same for the tsconfig.json under the api folder.  It is clear that your paths compilerOptions are not setup properly and not resolving the path from api/collections to api/server/collections.  This is certainly a source of confusion but makes for more simple import statements.  However, once you figure out that the "paths" compilerOptions allows you to throw you these curveballs you should be able to fix it.   In your main folder tsconfig.json make sure you have the "paths" set and set properly (NOTthe tsconfig.json in the api folder).  You main folder tsconfig.json should have the following "paths" compiler option:

    "paths": {
      "api/*": ["./api/server/*"]
    },

That is what is saying resolve every thing from api/* to ./api/server/* . So when you type 'api/collections' it is resolving to ./api/server/collections.  They did this to make the import statements more pretty and could also allow you to move things around without changing your page imports by changing the compilerOptions "paths".  Without using the "paths" option you could also do:

import { Chats } from '../../../api/server/collections';

But as you can see that is a bit like "walking the dog" trying to navigate to the proper folder.

Good luck and I am sure you will figure this out.

**Comment by sigalor** [comment]
Thanks a lot for your fast response, but as you can see [in my main tsconfig.json](https://github.com/sigalor/ionic-meteor-whatsapp-error/blob/master/tsconfig.json#L15), I already had the `paths` compiler option exactly as you told me to have it.
As said in the Typescript reference article about [Module Resolution](https://www.typescriptlang.org/docs/handbook/module-resolution.html#path-mapping), the `baseUrl` compiler option needs to be set to `"."`, [and it is](https://github.com/sigalor/ionic-meteor-whatsapp-error/blob/master/tsconfig.json#L4). But nonetheless, trying to specify `import { Chats } from "api/server/collections";` (having temporarily removed the 'paths' option in tsconfig.json) just fails with "Cannot find module" again.
This somehow has to be a bug in one of these compiler options, because when I try to use something like `import { Chats } from '../../../api/server/collections';` (without the 'paths' compiler option in the root tsconfig.json of course), everything works perfectly fine.

**Comment by clariontools** [comment]
That is about all I can offer on this, but at least now you can see it is an issue with the path resolution.  I did take a quick look at the main root tsconfig.json and noticed that you have the  "meteor-typings", in our compilerOptions "types" you should remove that as it conflicts with the meteor tsconfig.json that is found in your api folder.  Keep the paths option to resolve the api folder in your main tsconfig.json and you should be fine.

**Comment by sigalor** [comment]
Alright, I guess... So I think I'm going to have to keep the relative path for now. Hopefully, as I gain experience in the next few weeks/months, eventually I will be able to fix this by myself. But for now, thank you very much again for taking time to look into this.

**Comment by sigalor** [comment]
I finally managed to fix this! It really had to do with Webpack.

After the tutorial introduced Sharp, of course Webpack had to know that it must be executed on the server only. So, as the tutorial said, I added the line `{ sharp: "{}" }` to `webpack.config.js`'s exports. But this just did not work. After I created a syntax error in this js file (first accidentally, then intentionally), I understood what was happening: the `webpack.config.js` file was just ignored. So, after some more research, I found [this](https://stackoverflow.com/a/43429378) Stack Overflow answer. Although it did not really have anything to do with my issue, the solution was hidden in it. A bit embarassed, I further noticed that it was sitting [right there](https://github.com/Urigo/Ionic2CLI-Meteor-WhatsApp/blob/master/package.json#L92) in this repository the whole time.

TL;DR: Add the following lines to your `package.json`:

```
  "config": {
    "ionic_webpack": "./webpack.config.js",
  }
```

**Comment by clariontools** [comment]
Thanks for sharing, it does highlight how fragile this entire configuration is.  At least now if something breaks you have more knowledge of things to check.  :1st_place_medal:
